### PR TITLE
Hints: Separating out the range and offsets of PromQL subqueries

### LINKED ISSUE #7629: promql: Improve what is Selected when subquery is used.
As suggested by @gouthamve we can improve a lot on what we select form storage for subqueries.

See TODO: https://github.com/prometheus/prometheus/blob/ce1db029095e3246591fd157b3ccddce01313019/promql/engine.go#L654

### PR DESCRIPTION
**Pull request by roidelapluie**
Fix #7629

Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

File: promql/engine.go [hunk 1]
```diff
@@ -597,35 +597,39 @@ func (ng *Engine) execEvalStmt(ctx context.Context, query *query, s *parser.Eval
...
 	parser.Inspect(s.Expr, func(node parser.Node, path []parser.Node) error {
-		subqOffset := ng.cumulativeSubqueryOffset(path)
+		subqOffset, subqRange := ng.subqueryOffsetRange(path)
 		switch n := node.(type) {
 		case *parser.VectorSelector:
-			if maxOffset < ng.lookbackDelta+subqOffset {
-				maxOffset = ng.lookbackDelta + subqOffset
+			if maxOffset < ng.lookbackDelta+subqOffset+subqRange {
+				maxOffset = ng.lookbackDelta + subqOffset + subqRange
 			}
```
Comment (roidelapluie) [inline promql/engine.go:624]: @brian-brazil GIven than n.offset is > 0 always
those 3 lines above seem redundant with below?

File: promql/engine.go [hunk 2]
```diff
@@ -597,35 +597,39 @@ func (ng *Engine) execEvalStmt(ctx context.Context, query *query, s *parser.Eval
...
-				maxOffset = n.Offset + ng.lookbackDelta + subqOffset
+			if n.Offset+ng.lookbackDelta+subqOffset+subqRange > maxOffset {
+				maxOffset = n.Offset + ng.lookbackDelta + subqOffset + subqRange
 			}
 		case *parser.MatrixSelector:
-			if maxOffset < n.Range+subqOffset {
-				maxOffset = n.Range + subqOffset
+			if maxOffset < n.Range+subqOffset+subqRange {
+				maxOffset = n.Range + subqOffset + subqRange
 			}
```
Comment (roidelapluie) [inline promql/engine.go:631]: Same here

Comment (roidelapluie) [reply promql/engine.go:631]: Unless we are preparing for negatif offsets...

File: promql/engine.go (hunk 1 above)
Comment (bwplotka) [reply promql/engine.go:624]: +1, but happy with negative offset to be well handed here as well

**Comment by bwplotka** [comment]
I don't think https://github.com/prometheus/prometheus/issues/7629 is fixed by this, though. I think there is lots to improve in terms of `step` (interval). I think the lowest-level subquery interval should be passed as step.

Also it would be nice to add test cases for multiple nested subqueries within one query? :thinking:

**Comment by roidelapluie** [comment]
```
		query: "count_over_time((foo offset 10s)[2m:1s] offset 10s)", start: 300000, end: 500000,
￼		expected: []*storage.SelectHints{
￼			{Start: 155000, End: 480000, Func: "count_over_time", Step: **1000**},
￼		},
```
Sounds like that is the case

File: promql/engine.go (hunk 1 above)
Comment (brian-brazil) [reply promql/engine.go:624]: There can be multiple different offsets in one expression, which I believe this is handling. We may need a few more tests here.

Comment (roidelapluie) [reply promql/engine.go:603]: I have added a test. It handles foo offset 1m or foo offset 10m.

**Comment by roidelapluie** [comment]
@bwplotka I think it is all addressed. WDYT?

**Pull request merged**